### PR TITLE
Use AES256 for the default encryption algoritm for TLS session ticket…

### PR DESCRIPTION
**Pull request by Castaglia**
…s, rather

than AES128.  This involves providing more session ticket key data, for both
the cipher and the digest: 80 bytes total = key name(16) + cipher key(32) +
HMAC key(32).

This addresses Issue #514 .

### CODE REVIEW

**Comment by kroeckx** [comment]
This seems to be failing with:
*** glibc detected *** ./ssltest: free(): invalid pointer: 0x00000000027f609c ***

File: ssl/statem/statem_srvr.c [hunk 1]
```diff
@@ -3022,10 +3022,10 @@ int tls_construct_new_session_ticket(SSL *s)
     } else {
         if (RAND_bytes(iv, 16) <= 0)
             goto err;
-        if (!EVP_EncryptInit_ex(ctx, EVP_aes_128_cbc(), NULL,
+        if (!EVP_EncryptInit_ex(&ctx, EVP_aes_256_cbc(), NULL,
```
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: This & at least looks suspicious.

File: ssl/ssl_lib.c [hunk 2]
```diff
@@ -2335,10 +2335,10 @@ SSL_CTX *SSL_CTX_new(const SSL_METHOD *meth)
 
     ret->max_send_fragment = SSL3_RT_MAX_PLAIN_LENGTH;
 
-    /* Setup RFC4507 ticket keys */
+    /* Setup RFC5507 ticket keys */
```
Comment (kroeckx) [inline ssl/ssl_lib.c]: You probably mean RFC5077

File: ssl/s3_lib.c [hunk 3]
```diff
@@ -3524,19 +3524,19 @@ long ssl3_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
...
-                memcpy(ctx->tlsext_tick_hmac_key, keys + 16, 16);
-                memcpy(ctx->tlsext_tick_aes_key, keys + 32, 16);
+                memcpy(ctx->tlsext_tick_hmac_key, keys + 16, 32);
+                memcpy(ctx->tlsext_tick_aes_key, keys + 48, 32);
             } else {
                 memcpy(keys, ctx->tlsext_tick_key_name, 16);
-                memcpy(keys + 16, ctx->tlsext_tick_hmac_key, 16);
-                memcpy(keys + 32, ctx->tlsext_tick_aes_key, 16);
+                memcpy(keys + 16, ctx->tlsext_tick_hmac_key, 32);
+                memcpy(keys + 48, ctx->tlsext_tick_aes_key, 32);
```
Comment (kroeckx) [inline ssl/s3_lib.c]: Can you replace all those numbers like 16, 32, 48, 80 with something that makes sense, like sizeof(SSL_CTX_get_tlsext_ticket_keys)?

File: ssl/statem/statem_srvr.c (hunk 1 above)
Comment (Castaglia) [inline ssl/statem/statem_srvr.c]: Sorry about that.  Good catch, thanks!

File: ssl/ssl_lib.c (hunk 2 above)
Comment (Castaglia) [inline ssl/ssl_lib.c]: Indeed.  Fixed.

File: ssl/s3_lib.c (hunk 3 above)
Comment (Castaglia) [inline ssl/s3_lib.c]: Will do.  As a first pass, I'd wanted to keep in the same style as the existing code and get the approach reviewed, before refining further.

File: ssl/statem/statem_srvr.c (hunk 1 above)
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: You have the same problem in the other file.

File: ssl/statem/statem_srvr.c [hunk 4]
```diff
@@ -3022,10 +3022,10 @@ int tls_construct_new_session_ticket(SSL *s)
     } else {
         if (RAND_bytes(iv, 16) <= 0)
             goto err;
-        if (!EVP_EncryptInit_ex(ctx, EVP_aes_128_cbc(), NULL,
+        if (!EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), NULL,
                                 tctx->tlsext_tick_aes_key, iv))
             goto err;
-        if (!HMAC_Init_ex(hctx, tctx->tlsext_tick_hmac_key, 16,
+        if (!HMAC_Init_ex(hctx, tctx->tlsext_tick_hmac_key, 32,
```
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: That 32 should probably also change to sizeof(tctx->tlsext_tick_hmac_key)

File: ssl/t1_lib.c [hunk 5]
```diff
@@ -3095,15 +3095,17 @@ static int tls_decrypt_ticket(SSL *s, const unsigned char *etick,
             renew_ticket = 1;
     } else {
         /* Check key name matches */
-        if (memcmp(etick, tctx->tlsext_tick_key_name, 16)) {
+        if (memcmp(etick, tctx->tlsext_tick_key_name,
+                   sizeof(tctx->tlsext_tick_key_name))) {
```
Comment (kroeckx) [inline ssl/t1_lib.c]: Can you change to that != 0?

Comment (Castaglia) [inline ssl/t1_lib.c]: Done.

**Comment by ALSchwalm** [comment]
Is this still being worked on? I'm interested in seeing this (or any fix for #514) land. I think the problem travis is having at the moment is that it should be `sizeof(ctx->tlsext_tick_aes_key)` on line 3527.

**Comment by Castaglia** [comment]
I've updated the PR, and fixed that particular build issue (thanks!); it now appears that the Travis CI builds are failing for something unrelated to my changes.  Hopefully someone on the OpenSSL team could help diagnose those?

**Comment by kroeckx** [comment]
On Thu, Mar 17, 2016 at 07:12:19PM -0700, Adam Schwalm wrote:
> Is this still being worked on?

I've already moved it to the internal review some time ago.  I
think I've fixed at least 1 bug in it.

**Comment by kroeckx** [comment]
So this got merged.  Maybe you also want to create a merge request for 1.0.2?

**Comment by Castaglia** [comment]
@kroeckx done; see PR #1080 .  Thanks!

**Pull request closed without merging**